### PR TITLE
Level encoder should support writing levels for data page v2

### PR DESCRIPTION
**Pull request by sadikovi**
This PR attempts to update level encoder to support data pages v2 levels, i.e. starting with offset 0 instead of `mem::size_of<i32>()`.

When we write data page v1 levels, we encode the length of the levels in the first 4 bytes of the vector, hence `mem::size_of<i32>()` offset. With data page v2, we explicitly store the length of the levels, so we do not really need to those 4 bytes offset, and we would rather start with 0.

Level decoder already supports reading data page v1 and v2 by having two methods: `set_data`, which reads the first 4 bytes and sets the length, and `set_data_range` that allows to set start and length manually, since we know the start position and data page v2 header has the levels length.

I updated the relevant code and unit tests to reflect that.

### CODE REVIEW

**Comment by sadikovi** [comment]
@sunchao could you review this PR? I found this problem when writing tests for column writer. 

Fix works, but IMHO is a bit messy, and I think there is disconnection between level decoder and level encoder when it comes to encoding and decoding for data page v1 and data page v2. Let me know if it is okay, or we should refactor the code somehow. Thanks!

**Comment by sadikovi** [comment]
I think it might be better to have two level encoders and decoders for data page v1 and data page v2, it’s not that much of a code duplication and it makes it clear what functionality is supported for each.

**Comment by sunchao** [comment]
Thanks @sadikovi ! Yeah I'm not sure what's the most elegant way to handle this discrepancy between page v1 and v2. IMO we can just go with this approach and leave potential refactoring as a future work.

For reference, parquet-cpp doesn't seem to handle page v2. On the other hand, parquet-mr doesn't have a level encoder/decoder. Instead it (taking decoding path as example, encoding should be symmetric):

1. if it's v1, initiate value reader based on encoding (RLE or BIT-PACKED), and the RLE (in our case `RleValueDecoder`) will handle the 4-byte length. In our case, we don't have a BIT-PACKED value reader since this is only used for decoding plain boolean values and our plain decoder already does that. BTW: I think #61 is not an issue anymore.

2. if it's v2, initiate a "raw" RLE value decoder, and just pass the definition/repetition levels byte stream to it (unlike our page structure, parquet-mr has separate byte streams for def/rep levels for page v2).

IMO I don't see a clear advantage in either approach.

**Comment by sadikovi** [comment]
Let me try separating v1 from v2 and see if it makes things simpler.

**Comment by sadikovi** [comment]
@sunchao I updated the code to see if it is better to introduce `RLE_V2` for Data Page v2, also did some minor refactoring. Let me know if this approach is better, or if I should revert to the previous approach. Thanks!

File: src/encodings/levels.rs [hunk 1]
```diff
@@ -25,52 +25,70 @@ use errors::{ParquetError, Result};
...
+  }
 }
 
-/// A encoder for definition/repetition levels.
-/// Currently only supports RLE and BIT_PACKED (dev/null) encoding.
-pub struct LevelEncoder {
-  bit_width: u8,
-  encoder: InternalEncoder
+/// Encoder for definition/repetition levels.
+/// Currently only supports RLE and BIT_PACKED (dev/null) encoding, including v2.
```
Comment (sunchao) [inline src/encodings/levels.rs]: what does `dev/null` here mean?

File: src/encodings/levels.rs [hunk 2]
```diff
@@ -25,52 +25,70 @@ use errors::{ParquetError, Result};
...
+pub enum LevelEncoder {
+  RLE(RleEncoder),
+  BIT_PACKED(u8, BitWriter),
+  RLE_V2(RleEncoder)
```
Comment (sunchao) [inline src/encodings/levels.rs]: nit nit: can we move `RLE_V2` next to `RLE`?

File: src/encodings/levels.rs [hunk 3]
```diff
@@ -25,52 +25,70 @@ use errors::{ParquetError, Result};
...
 }
 
 impl LevelEncoder {
   /// Creates new level encoder based on encoding, max level and underlying byte buffer.
   /// For bit packed encoding it is assumed that buffer is already allocated with
-  /// 'LevelEncoder::max_buffer_size' method.
+  /// 'levels::max_buffer_size' method.
```
Comment (sunchao) [inline src/encodings/levels.rs]: nit: can we use `levels::max_buffer_size` here (the backtick syntax).

File: src/encodings/levels.rs (hunk 1 above)
Comment (sadikovi) [reply src/encodings/levels.rs]: I do not quite remember, it was something similar to no-op encoding. I will remove it.

**Comment by sunchao** [comment]
Looks good @sadikovi . Will merge shortly after the CI.

**Comment by sadikovi** [comment]
Thanks!

**Comment by sadikovi** [comment]
Thanks for merging!

**Pull request merged**